File: runme/mcgen/local/datan/solsrc/s1reg.f
      PROGRAM S1REG
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=100,NRMAX=10,NR=10,NPL=200)
      DIMENSION T(NMAX),Y(NMAX),DELTAY(NMAX),X(NRMAX),CHI2(NRMAX)
      DIMENSION B(NRMAX,NRMAX),A(NMAX,NRMAX)
      DIMENSION XPL(NPL,NR),YPL(NPL,NR),NCOL(NR)
      DIMENSION DATSX(NMAX),DATCOV(NMAX)
C COMMON /CS1REG/ is needed for Microsoft FORTRAN compiler
      COMMON /CS1REG/ T,Y,DELTAY,X,CHI2,B,A,XPL,YPL,DATSX,DATCOV
      CHARACTER*75 TX,TY,CAPT
      PARAMETER (ZERO=0.D0,ONE=1.D0)
C identify program to user
      WRITE(*,*)' Program S1REG simulates data points with errors,'
      WRITE(*,*)' performs polynomial regression on them'
      WRITE(*,*)' and presents data and results graphically'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number N of data points (1 <= N <= 100)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter SIGMA (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)
     +' Enter number M of terms in polynomial for data (1<=M<=N)'
      WRITE(*,*)' >'
      READ(*,*)M
      DO 2 I=1,M
        WRITE(*,'(A,I3,A)')' Enter X(',I,' ) for simulation'
        WRITE(*,*)' >'
        READ(*,*)X(I)
2     CONTINUE
      T0=-1.D0
      DT=2.D0/DBLE(N-1)
C store random numbers following standardized normal in Y
      CALL RNSTNR(Y,N)
C loop over N data points
      DO 10 I=1,N
        T(I)=T0+(I-1)*DT
          DAT=X(1)
        IF(M.GT.1) THEN
          DO 5 J=2,M
            DAT=DAT+X(J)*T(I)**(J-1)
5         CONTINUE
        END IF
C Now DAT is exactly given by polynomial.
C Add normal error of width SIGMA and store result in Y
        Y(I)=DAT+Y(I)*SIGMA
C Set error of data point equal to SIGMA
        DELTAY(I)=SIGMA
10    CONTINUE
C Perform polynomial regression
      CALL REGPOL(T,Y,DELTAY,N,NR,X,B,A,CHI2)
C Output results
      WRITE(*,'(A)')' T '
      WRITE(*,'(1X,10F7.2)')(T(I),I=1,N)
      WRITE(*,'(A)')' Y'
      WRITE(*,'(1X,10F7.2)')(Y(I),I=1,N)
      WRITE(*,'(A)')' DELTAY'
      WRITE(*,'(1X,10F7.2)')(DELTAY(I),I=1,N)
      WRITE(*,'(A)')' X'
      WRITE(*,'(1X,10F7.2)')(X(I),I=1,NR)
      WRITE(*,'(A)')' CHI2'
      WRITE(*,'(1X,10F7.2)')(CHI2(I),I=1,NR)
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Enter maximum number of terms in polynomials'
      WRITE(*,*)' to be plotted (<= 10)'
      WRITE(*,*)' >'
      READ(*,*)NPOL
C set errors along abscissa and covariances to zero for graphics
      DO 15 I=1,N
        DATSX(I)=0.D0
        DATCOV(I)=0.D0
15    CONTINUE
C compute polylines for polynomials of degrees 0,1,2,....
      T0=-1.5D0
      DT=3.D0/DBLE(NPL-1)
      DO 40 I=1,NPL
        DO 30 J=1,NPOL
          XPL(I,J)=T0+(I-1)*DT
          YPL(I,J)=ZERO
          DO 27 L=1,J
            D=B(L,1)
            IF(L.GT.1) THEN
              DO 25 K=2,L
                D=D+B(L,K)*XPL(I,J)**(K-1)
25            CONTINUE
            END IF
            YPL(I,J)=YPL(I,J)+X(L)*D
27        CONTINUE
30      CONTINUE
40    CONTINUE
      DO 50 J=1,10
        NCOL(J)=4
50    CONTINUE
C prepare texts for graphics
      TX='t'
      LTX=1
      TY='y'
      LTY=1
      CAPT=' '
      LCAPT=1
      NMARK=5
      SCALEF=.5D0
C call graphics routine for data points and multiple polylines
      CALL GRDTMC(XPL,YPL,NPL,NPOL,NCOL,NMARK,SCALEF, T,Y,DATSX,
     +DELTAY,DATCOV,N,TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END

File: runme/mcgen/local/datan/solsrc/s1reg.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Common Block Declarations */
/* COMMON /CS1REG/ is needed for Microsoft FORTRAN compiler */
struct {
    double t[100], y[100], deltay[100], x[10], chi2[10],
           b[100] /* was [10][10] */, *a /* was [100][10] */,
           *xpl /* was [200][10] */, *ypl /* was [200][10] */,
           datsx[100], datcov[100];
} cs1reg;

/* Table of constant values */

static integer c__10 = 10, c__200 = 200;

main(void)
{
    /* Local variables */
    static integer i, j, k, l, m, n, lcapt, nmark, ltx, lty, nws,
                   ncol[10], npol;
    static double d, dat, sigma, t0, dt, scalef;
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1, i__2, i__3, i__4;

    if ((cs1reg.a = (double *)MEMALLOC(sizeof(double)*5000)) == NULL) {
      MEMERROR("s1reg");
      return(0);
    } else {
      cs1reg.xpl = cs1reg.a+1000;
      cs1reg.ypl = cs1reg.xpl+2000;
    }
/* identify program to user */
    printf("%s\n", "Program S1REG simulates data points with errors,");
    printf("%s\n", "performs polynomial regression on them");
    printf("%s\n\n", "and presents data and results graphically.");
/* ask for input */
    printf("%s\n", "Enter number N of data points (1 <= N <= 100):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter SIGMA (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Enter number M of terms in polynomial for data (1<=M<=N):");
    printf("%s", "> ");
    scanf("%li", &m);
    i__1 = m;
    for (i = 1; i <= i__1; ++i) {
        printf("%s%3li%s\n", "Enter X(", i, " ) for simulation:");
        printf("%s", "> ");
        scanf("%lf", &cs1reg.x[i-1]);
    }
    t0 = -1.;
    dt = 2. / (double) (n - 1);
/* store random numbers following standardized normal in Y */
    rnstnr_(cs1reg.y, &n);
/* loop over N data points */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        cs1reg.t[i - 1] = t0 + (i - 1) * dt;
        dat = cs1reg.x[0];
        if (m > 1) {
            i__2 = m;
            for (j = 2; j <= i__2; ++j) {
                i__3 = j - 1;
                dat += cs1reg.x[j - 1] * pow(cs1reg.t[i - 1], (double)i__3);
            }
        }
/* Now DAT is exactly given by polynomial. */
/* Add normal error of width SIGMA and store result in Y */
        cs1reg.y[i - 1] = dat + cs1reg.y[i - 1] * sigma;
/* Set error of data point equal to SIGMA */
        cs1reg.deltay[i - 1] = sigma;
    }
/* Perform polynomial regression */
    regpol_(cs1reg.t, cs1reg.y, cs1reg.deltay, &n, &c__10, cs1reg.x,
            cs1reg.b, cs1reg.a, cs1reg.chi2);
/* Output results */
    printf("%s\n", "T");
    for (i = 1; i <= n; ++i) {
        if (i % 10 == 1)
            printf(" ");
        printf("%7.2lf", cs1reg.t[i-1]);
        if (i % 10 == 0)
           printf("\n");
    }
    if (n % 10 != 0)
       printf("\n");
    printf("%s\n", "Y");
    for (i = 1; i <= n; ++i) {
        if (i % 10 == 1)
            printf(" ");
        printf("%7.2lf", cs1reg.y[i-1]);
        if (i % 10 == 0)
           printf("\n");
    }
    if (n % 10 != 0)
       printf("\n");
    printf("%s\n", "DELTAY");
    for (i = 1; i <= n; ++i) {
        if (i % 10 == 1)
            printf(" ");
        printf("%7.2lf", cs1reg.deltay[i-1]);
        if (i % 10 == 0)
           printf("\n");
    }
    if (n % 10 != 0)
       printf("\n");
    printf("%s\n ", "X");
    for (i = 1; i <= 10; ++i) {
        printf("%7.2lf", cs1reg.x[i-1]);
    }
    printf("\n");
    printf("%s\n ", "CHI2");
    for (i = 1; i <= 10; ++i) {
        printf("%7.2lf", cs1reg.chi2[i-1]);
    }
    printf("\n");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Enter maximum number of terms in polynomials");
    printf("%s\n", "to be plotted (<= 10):");
    printf("%s", "> ");
    scanf("%li", &npol);
    (void)getchar();
/* set errors along abscissa and covariances to zero for graphics */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        cs1reg.datsx[i - 1] = 0.;
        cs1reg.datcov[i - 1] = 0.;
    }
/* compute polylines for polynomials of degrees 0,1,2,.... */
    t0 = -1.5;
    dt = .01507537688442211;
    for (i = 1; i <= 200; ++i) {
        i__1 = npol;
        for (j = 1; j <= i__1; ++j) {
            cs1reg.xpl[i + j * 200 - 201] = t0 + (i - 1) * dt;
            cs1reg.ypl[i + j * 200 - 201] = 0.;
            i__2 = j;
            for (l = 1; l <= i__2; ++l) {
                d = cs1reg.b[l - 1];
                if (l > 1) {
                    i__3 = l;
                    for (k = 2; k <= i__3; ++k) {
                        i__4 = k - 1;
                        d += cs1reg.b[l + k * 10 - 11] *
                             pow(cs1reg.xpl[i + j * 200 - 201], (double)i__4);
                    }
                }
                cs1reg.ypl[i + j * 200 - 201] += cs1reg.x[l - 1] * d;
            }
        }
    }
    for (j = 1; j <= 10; ++j) {
        ncol[j - 1] = 4;
    }
/* prepare texts for graphics */
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    strncpy(tx, "t", 1);
    ltx = 1;
    strncpy(ty, "y", 1);
    lty = 1;
    strncpy(capt, " ", 1);
    lcapt = 1;
    nmark = 5;
    scalef = .5;
/* call graphics routine for data points and multiple polylines */
    grdtmc_(cs1reg.xpl, cs1reg.ypl, &c__200, &npol, ncol, &nmark, &scalef,
           cs1reg.t, cs1reg.y, cs1reg.datsx, cs1reg.deltay,
           cs1reg.datcov, &n, tx, &ltx, ty, &lty, capt, &lcapt, &nws);
    MEMFREE((void *) cs1reg.a);
    return 0;
} /* main */

